### PR TITLE
Conceal the full type of the "Splash" from the window management policy

### PR DESCRIPTION
**Pull request by AlanGriffiths**
Conceal the full type of the "Splash" from the window management policy.

This will make it possible to substitute different "Splash" implementations. In particular, I want to offer an example Wayland based "Splash".

### CODE REVIEW

File: examples/example-server-lib/splash_session.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    struct Self;
+    SplashSession(std::shared_ptr<Self> const self);
+
+    auto session() const -> std::shared_ptr<mir::scene::Session>;
+
+private:
+    std::shared_ptr<Self> const self;
+};
+
+class SplashSession::Self
```
Comment (wmww) [inline examples/example-server-lib/splash_session.h]: are we using PImpl here, especially when the implementation is in the same file?

Comment (RAOF) [reply examples/example-server-lib/splash_session.h]: The reason to have separate `SplashSession` and `SplashSession::Self` types here is ABI stability, yes?

You can later add a `SplashSession::SelfV2` struct with more virtual functions and `SplashSession` can `dynamic_cast` its way to using them.

**Comment by AlanGriffiths** [comment]
This isn't about ABI (it is just sample code) the point is type erasure: the users of SplashSession (i.e. any window managers) don't need to care about the actual type. See the use I make of this in #421.

**Comment by wmww** [comment]
What's the reason for not combining the two classes like this: https://github.com/MirServer/mir/commit/9f33948955719d4fccc8c56bd22f021b6a8b57e9? Window managers knowing the type doens't do any harm if they need to be recompiled when it changes anyway.

**Comment by AlanGriffiths** [comment]
I guess our views on complexity differ. To me passing around a `Foo` is simpler that passing around a `std::shared_ptr<Bar>`.

**Comment by RAOF** [comment]
I think I'm more in line with William's thinking here - we don't have a lot of handle-type wrappers in the code at the moment, and I think that passing `std::shared_ptr<Bar>` around is no less simple and is more clear than passing around a `Foo` that happens to behave like a `std::shared_ptr`.

Particularly, WRT to #421, it seems odd to me that you're implementing `SpinnerSplash` IS-A `SplashSession` via `operator SplashSession()` rather than implementing an interface.

I'm not strongly opposed, though, particularly since the code is already written :grin:. Feel free to either rework it as William suggested, or land it as-is.

bors delegate+

**Comment by wmww** [comment]
oops

**Comment by AlanGriffiths** [comment]
Closing and restoring seems a neat way to re-run the Travis build...

**Comment by AlanGriffiths** [comment]
...and we've had enough hours to wonder if CI is fixed.

**Comment by AlanGriffiths** [comment]
bors r=RAOF.wmww

**Pull request merged**